### ISSUE TITLE
PEP 498: Formatted string literals

### ISSUE THREAD

**Issue by anijjar** [anijjar/ironpython3#1]
https://docs.python.org/3.6/whatsnew/3.6.html#pep-498-formatted-string-literals

https://www.python.org/dev/peps/pep-0498

https://docs.python.org/3.6/reference/lexical_analysis.html#f-strings

**Comment by anijjar** [comment]
PEP: 498
Title: Literal String Interpolation
Version: $Revision$
Last-Modified: $Date$
Author: Eric V. Smith <eric@trueblade.com>
Status: Final
Type: Standards Track
Content-Type: text/x-rst
Created: 01-Aug-2015
Python-Version: 3.6
Post-History: 07-Aug-2015, 30-Aug-2015, 04-Sep-2015, 19-Sep-2015, 06-Nov-2016
Resolution: https://mail.python.org/pipermail/python-dev/2015-September/141526.html

Abstract
========

Python supports multiple ways to format text strings. These include
%-formatting [#]_, ``str.format()`` [#]_, and ``string.Template``
[#]_. Each of these methods have their advantages, but in addition
have disadvantages that make them cumbersome to use in practice. This
PEP proposed to add a new string formatting mechanism: Literal String
Interpolation. In this PEP, such strings will be referred to as
"f-strings", taken from the leading character used to denote such
strings, and standing for "formatted strings".

This PEP does not propose to remove or deprecate any of the existing
string formatting mechanisms.

F-strings provide a way to embed expressions inside string literals,
using a minimal syntax. It should be noted that an f-string is really
an expression evaluated at run time, not a constant value. In Python
source code, an f-string is a literal string, prefixed with 'f', which
contains expressions inside braces. The expressions are replaced with
their values. Some examples are::

  >>> import datetime
  >>> name = 'Fred'
  >>> age = 50
  >>> anniversary = datetime.date(1991, 10, 12)
  >>> f'My name is {name}, my age next year is {age+1}, my anniversary is {anniversary:%A, %B %d, %Y}.'
  'My name is Fred, my age next year is 51, my anniversary is Saturday, October 12, 1991.'
  >>> f'He said his name is {name!r}.'
  "He said his name is 'Fred'."

A similar feature was proposed in PEP 215. PEP 215 proposed to support
a subset of Python expressions, and did not support the type-specific
string formatting (the ``__format__()`` method) which was introduced
with PEP 3101.

Rationale
=========

This PEP is driven by the desire to have a simpler way to format
strings in Python. The existing ways of formatting are either error
prone, inflexible, or cumbersome.

%-formatting is limited as to the types it supports. Only ints, strs,
and doubles can be formatted. All other types are either not
supported, or converted to one of these types before formatting. In
addition, there's a well-known trap where a single value is passed::

  >>> msg = 'disk failure'
  >>> 'error: %s' % msg
  'error: disk failure'

But if msg were ever to be a tuple, the same code would fail::

  >>> msg = ('disk failure', 32)
  >>> 'error: %s' % msg
  Traceback (most recent call last):
    File "<stdin>", line 1, in <module>
  TypeError: not all arguments converted during string formatting

To be defensive, the following code should be used::

  >>> 'error: %s' % (msg,)
  "error: ('disk failure', 32)"

``str.format()`` was added to address some of these problems with
%-formatting. In particular, it uses normal function call syntax (and
therefor supports multiple parameters) and it is extensible through
the ``__format__()`` method on the object being converted to a
string. See PEP 3101 for a detailed rationale. This PEP reuses much of
the ``str.format()`` syntax and machinery, in order to provide
continuity with an existing Python string formatting mechanism.

However, ``str.format()`` is not without its issues. Chief among them
is its verbosity. For example, the text ``value`` is repeated here::

  >>> value = 4 * 20
  >>> 'The value is {value}.'.format(value=value)
  'The value is 80.'

Even in its simplest form there is a bit of boilerplate, and the value
that's inserted into the placeholder is sometimes far removed from
where the placeholder is situated::

  >>> 'The value is {}.'.format(value)
  'The value is 80.'

With an f-string, this becomes::

  >>> f'The value is {value}.'
  'The value is 80.'

F-strings provide a concise, readable way to include the value of
Python expressions inside strings.

In this sense, ``string.Template`` and %-formatting have similar
shortcomings to ``str.format()``, but also support fewer formatting
options. In particular, they do not support the ``__format__``
protocol, so that there is no way to control how a specific object is
converted to a string, nor can it be extended to additional types that
want to control how they are converted to strings (such as ``Decimal``
and ``datetime``). This example is not possible with
``string.Template``::

  >>> value = 1234
  >>> f'input={value:#06x}'
  'input=0x04d2'

And neither %-formatting nor ``string.Template`` can control
formatting such as::

  >>> date = datetime.date(1991, 10, 12)
  >>> f'{date} was on a {date:%A}'
  '1991-10-12 was on a Saturday'

No use of globals() or locals()
-------------------------------

In the discussions on python-dev [#]_, a number of solutions where
presented that used locals() and globals() or their equivalents. All
of these have various problems. Among these are referencing variables
that are not otherwise used in a closure. Consider::

  >>> def outer(x):
  ...     def inner():
  ...         return 'x={x}'.format_map(locals())
  ...     return inner
  ...
  >>> outer(42)()
  Traceback (most recent call last):
    File "<stdin>", line 1, in <module>
    File "<stdin>", line 3, in inner
  KeyError: 'x'

This returns an error because the compiler has not added a reference
to x inside the closure. You need to manually add a reference to x in
order for this to work::

  >>> def outer(x):
  ...     def inner():
  ...         x
  ...         return 'x={x}'.format_map(locals())
  ...     return inner
  ...
  >>> outer(42)()
  'x=42'

In addition, using locals() or globals() introduces an information
leak. A called routine that has access to the callers locals() or
globals() has access to far more information than needed to do the
string interpolation.

Guido stated [#]_ that any solution to better string interpolation
would not use locals() or globals() in its implementation.  (This does
not forbid users from passing locals() or globals() in, it just
doesn't require it, nor does it allow using these functions under the
hood.)

Specification
=============

In source code, f-strings are string literals that are prefixed by the
letter 'f' or 'F'. Everywhere this PEP uses 'f', 'F' may also be
used. 'f' may be combined with 'r' or 'R', in either order, to produce
raw f-string literals. 'f' may not be combined with 'b': this PEP does
not propose to add binary f-strings. 'f' may not be combined with 'u'.

When tokenizing source files, f-strings use the same rules as normal
strings, raw strings, binary strings, and triple quoted strings. That
is, the string must end with the same character that it started with:
if it starts with a single quote it must end with a single quote, etc.
This implies that any code that currently scans Python code looking
for strings should be trivially modifiable to recognize f-strings
(parsing within an f-string is another matter, of course).

Once tokenized, f-strings are parsed in to literal strings and
expressions.  Expressions appear within curly braces ``'{'`` and
``'}'``.  While scanning the string for expressions, any doubled
braces ``'{{'`` or ``'}}'`` inside literal portions of an f-string are
replaced by the corresponding single brace.  Doubled literal opening
braces do not signify the start of an expression.  A single closing
curly brace ``'}'`` in the literal portion of a string is an error:
literal closing curly braces must be doubled ``'}}'`` in order to
represent a single closing brace.

The parts of the f-string outside of braces are literal
strings.  These literal portions are then decoded.  For non-raw
f-strings, this includes converting backslash escapes such as
``'\n'``, ``'\"'``, ``"\'"``, ``'\xhh'``, ``'\uxxxx'``,
``'\Uxxxxxxxx'``, and named unicode characters ``'\N{name}'`` into
their associated Unicode characters [#]_.

Backslashes may not appear anywhere within expressions.  Comments,
using the ``'#'`` character, are not allowed inside an expression.

Following each expression, an optional type conversion may be
specified.  The allowed conversions are ``'!s'``, ``'!r'``, or
``'!a'``. These are treated the same as in ``str.format()``: ``'!s'``
calls ``str()`` on the expression, ``'!r'`` calls ``repr()`` on the
expression, and ``'!a'`` calls ``ascii()`` on the expression. These
conversions are applied before the call to ``format()``. The only
reason to use ``'!s'`` is if you want to specify a format specifier
that applies to ``str``, not to the type of the expression.

F-strings use the same format specifier mini-language as ``str.format``.
Similar to ``str.format()``, optional format specifiers maybe be
included inside the f-string, separated from the expression (or the
type conversion, if specified) by a colon. If a format specifier is
not provided, an empty string is used.

So, an f-string looks like::

  f ' <text> { <expression> <optional !s, !r, or !a> <optional : format specifier> } <text> ... '

The expression is then formatted using the ``__format__`` protocol,
using the format specifier as an argument. The resulting value is
used when building the value of the f-string.

Note that ``__format__()`` is not called directly on each value. The
actual code uses the equivalent of ``type(value).__format__(value,
format_spec)``, or ``format(value, format_spec)``. See the
documentation of the builtin ``format()`` function for more details.

Expressions cannot contain ``':'`` or ``'!'`` outside of strings or
parentheses, brackets, or braces. The exception is that the ``'!='``
operator is allowed as a special case.

Escape sequences
----------------

Backslashes may not appear inside the expression portions of
f-strings, so you cannot use them, for example, to escape quotes
inside f-strings::

  >>> f'{\'quoted string\'}'
    File "<stdin>", line 1
  SyntaxError: f-string expression part cannot include a backslash

You can use a different type of quote inside the expression::

  >>> f'{"quoted string"}'
  'quoted string'

Backslash escapes may appear inside the string portions of an
f-string.

Note that the correct way to have a literal brace appear in the
resulting string value is to double the brace::

  >>> f'{{ {4*10} }}'
  '{ 40 }'
  >>> f'{{{4*10}}}'
  '{40}'

Like all raw strings in Python, no escape processing is done for raw
f-strings::

  >>> fr'x={4*10}\n'
  'x=40\\n'

Due to Python's string tokenizing rules, the f-string
``f'abc {a['x']} def'`` is invalid. The tokenizer parses this as 3
tokens: ``f'abc {a['``, ``x``, and ``']} def'``. Just like regular
strings, this cannot be fixed by using raw strings. There are a number
of correct ways to write this f-string: with a different quote
character::

  f"abc {a['x']} def"

Or with triple quotes::

  f'''abc {a['x']} def'''

Code equivalence
----------------

The exact code used to implement f-strings is not specified. However,
it is guaranteed that any embedded value that is converted to a string
will use that value's ``__format__`` method. This is the same
mechanism that ``str.format()`` uses to convert values to strings.

For example, this code::

  f'abc{expr1:spec1}{expr2!r:spec2}def{expr3}ghi'

Might be evaluated as::

  'abc' + format(expr1, spec1) + format(repr(expr2), spec2) + 'def' + format(expr3) + 'ghi'

Expression evaluation
---------------------

The expressions that are extracted from the string are evaluated in
the context where the f-string appeared. This means the expression has
full access to local and global variables. Any valid Python expression
can be used, including function and method calls.

Because the f-strings are evaluated where the string appears in the
source code, there is no additional expressiveness available with
f-strings. There are also no additional security concerns: you could
have also just written the same expression, not inside of an
f-string::

  >>> def foo():
  ...   return 20
  ...
  >>> f'result={foo()}'
  'result=20'

Is equivalent to::

  >>> 'result=' + str(foo())
  'result=20'

Expressions are parsed with the equivalent of ``ast.parse('(' +
expression + ')', '<fstring>', 'eval')`` [#]_.

Note that since the expression is enclosed by implicit parentheses
before evaluation, expressions can contain newlines. For example::

  >>> x = 0
  >>> f'''{x
  ... +1}'''
  '1'

  >>> d = {0: 'zero'}
  >>> f'''{d[0
  ... ]}'''
  'zero'

Format specifiers
-----------------

Format specifiers may also contain evaluated expressions. This allows
code such as::

  >>> width = 10
  >>> precision = 4
  >>> value = decimal.Decimal('12.34567')
  >>> f'result: {value:{width}.{precision}}'
  'result:      12.35'

Once expressions in a format specifier are evaluated (if necessary),
format specifiers are not interpreted by the f-string evaluator. Just
as in ``str.format()``, they are merely passed in to the
``__format__()`` method of the object being formatted.

Concatenating strings
---------------------

Adjacent f-strings and regular strings are concatenated. Regular
strings are concatenated at compile time, and f-strings are
concatenated at run time. For example, the expression::

  >>> x = 10
  >>> y = 'hi'
  >>> 'a' 'b' f'{x}' '{c}' f'str<{y:^4}>' 'd' 'e'

yields the value::

  'ab10{c}str< hi >de'

While the exact method of this run time concatenation is unspecified,
the above code might evaluate to::

  'ab' + format(x) + '{c}' + 'str<' + format(y, '^4') + '>de'

Each f-string is entirely evaluated before being concatenated to
adjacent f-strings. That means that this::

  >>> f'{x' f'}'

Is a syntax error, because the first f-string does not contain a
closing brace.

Error handling
--------------

Either compile time or run time errors can occur when processing
f-strings. Compile time errors are limited to those errors that can be
detected when scanning an f-string. These errors all raise
``SyntaxError``.

Unmatched braces::

  >>> f'x={x'
    File "<stdin>", line 1
  SyntaxError: f-string: expecting '}'

Invalid expressions::

  >>> f'x={!x}'
    File "<stdin>", line 1
  SyntaxError: f-string: empty expression not allowed

Run time errors occur when evaluating the expressions inside an
f-string. Note that an f-string can be evaluated multiple times, and
work sometimes and raise an error at other times::

  >>> d = {0:10, 1:20}
  >>> for i in range(3):
  ...     print(f'{i}:{d[i]}')
  ...
  0:10
  1:20
  Traceback (most recent call last):
    File "<stdin>", line 2, in <module>
  KeyError: 2

or::

  >>> for x in (32, 100, 'fifty'):
  ...   print(f'x = {x:+3}')
  ...
  'x = +32'
  'x = +100'
  Traceback (most recent call last):
    File "<stdin>", line 2, in <module>
  ValueError: Sign not allowed in string format specifier

Leading and trailing whitespace in expressions is ignored
---------------------------------------------------------

For ease of readability, leading and trailing whitespace in
expressions is ignored. This is a by-product of enclosing the
expression in parentheses before evaluation.

Evaluation order of expressions
-------------------------------

The expressions in an f-string are evaluated in left-to-right
order. This is detectable only if the expressions have side effects::

  >>> def fn(l, incr):
  ...    result = l[0]
  ...    l[0] += incr
  ...    return result
  ...
  >>> lst = [0]
  >>> f'{fn(lst,2)} {fn(lst,3)}'
  '0 2'
  >>> f'{fn(lst,2)} {fn(lst,3)}'
  '5 7'
  >>> lst
  [10]

Discussion
==========

python-ideas discussion
-----------------------

Most of the discussions on python-ideas [#]_ focused on three issues:

- How to denote f-strings,
- How to specify the location of expressions in f-strings, and
- Whether to allow full Python expressions.

How to denote f-strings
***********************

Because the compiler must be involved in evaluating the expressions
contained in the interpolated strings, there must be some way to
denote to the compiler which strings should be evaluated. This PEP
chose a leading ``'f'`` character preceding the string literal. This
is similar to how ``'b'`` and ``'r'`` prefixes change the meaning of
the string itself, at compile time. Other prefixes were suggested,
such as ``'i'``. No option seemed better than the other, so ``'f'``
was chosen.

Another option was to support special functions, known to the
compiler, such as ``Format()``. This seems like too much magic for
Python: not only is there a chance for collision with existing
identifiers, the PEP author feels that it's better to signify the
magic with a string prefix character.

How to specify the location of expressions in f-strings
*******************************************************

This PEP supports the same syntax as ``str.format()`` for
distinguishing replacement text inside strings: expressions are
contained inside braces. There were other options suggested, such as
``string.Template``'s ``$identifier`` or ``${expression}``.

While ``$identifier`` is no doubt more familiar to shell scripters and
users of some other languages, in Python ``str.format()`` is heavily
used. A quick search of Python's standard library shows only a handful
of uses of ``string.Template``, but hundreds of uses of
``str.format()``.

Another proposed alternative was to have the substituted text between
``\{`` and ``}`` or between ``\{`` and ``\}``. While this syntax would
probably be desirable if all string literals were to support
interpolation, this PEP only supports strings that are already marked
with the leading ``'f'``. As such, the PEP is using unadorned braces
to denoted substituted text, in order to leverage end user familiarity
with ``str.format()``.

Supporting full Python expressions
**********************************

Many people on the python-ideas discussion wanted support for either
only single identifiers, or a limited subset of Python expressions
(such as the subset supported by ``str.format()``). This PEP supports
full Python expressions inside the braces. Without full expressions,
some desirable usage would be cumbersome. For example::

  >>> f'Column={col_idx+1}'
  >>> f'number of items: {len(items)}'

would become::

  >>> col_number = col_idx+1
  >>> f'Column={col_number}'
  >>> n_items = len(items)
  >>> f'number of items: {n_items}'

While it's true that very ugly expressions could be included in the
f-strings, this PEP takes the position that such uses should be
addressed in a linter or code review::

  >>> f'mapping is { {a:b for (a, b) in ((1, 2), (3, 4))} }'
  'mapping is {1: 2, 3: 4}'

Similar support in other languages
----------------------------------

Wikipedia has a good discussion of string interpolation in other
programming languages [#]_. This feature is implemented in many
languages, with a variety of syntaxes and restrictions.

Differences between f-string and str.format expressions
-------------------------------------------------------

There is one small difference between the limited expressions allowed
in ``str.format()`` and the full expressions allowed inside
f-strings. The difference is in how index lookups are performed. In
``str.format()``, index values that do not look like numbers are
converted to strings::

  >>> d = {'a': 10, 'b': 20}
  >>> 'a={d[a]}'.format(d=d)
  'a=10'

Notice that the index value is converted to the string ``'a'`` when it
is looked up in the dict.

However, in f-strings, you would need to use a literal for the value
of ``'a'``::

  >>> f'a={d["a"]}'
  'a=10'

This difference is required because otherwise you would not be able to
use variables as index values::

  >>> a = 'b'
  >>> f'a={d[a]}'
  'a=20'

See [#]_ for a further discussion. It was this observation that led to
full Python expressions being supported in f-strings.

Furthermore, the limited expressions that ``str.format()`` understands
need not be valid Python expressions. For example::

  >>> '{i[";]}'.format(i={'";':4})
  '4'

For this reason, the str.format() "expression parser" is not suitable
for use when implementing f-strings.

Triple-quoted f-strings
-----------------------

Triple quoted f-strings are allowed. These strings are parsed just as
normal triple-quoted strings are. After parsing and decoding, the
normal f-string logic is applied, and ``__format__()`` is called on
each value.

Raw f-strings
-------------

Raw and f-strings may be combined. For example, they could be used to
build up regular expressions::

  >>> header = 'Subject'
  >>> fr'{header}:\s+'
  'Subject:\\s+'

In addition, raw f-strings may be combined with triple-quoted strings.

No binary f-strings
-------------------

For the same reason that we don't support ``bytes.format()``, you may
not combine ``'f'`` with ``'b'`` string literals. The primary problem
is that an object's ``__format__()`` method may return Unicode data that
is not compatible with a bytes string.

Binary f-strings would first require a solution for
``bytes.format()``. This idea has been proposed in the past, most
recently in PEP 461 [#]_. The discussions of such a feature usually
suggest either

- adding a method such as ``__bformat__()`` so an object can control
  how it is converted to bytes, or

- having ``bytes.format()`` not be as general purpose or extensible
  as ``str.format()``.

Both of these remain as options in the future, if such functionality
is desired.

``!s``, ``!r``, and ``!a`` are redundant
----------------------------------------

The ``!s``, ``!r``, and ``!a`` conversions are not strictly
required. Because arbitrary expressions are allowed inside the
f-strings, this code::

  >>> a = 'some string'
  >>> f'{a!r}'
  "'some string'"

Is identical to::

  >>> f'{repr(a)}'
  "'some string'"

Similarly, ``!s`` can be replaced by calls to ``str()`` and ``!a`` by
calls to ``ascii()``.

However, ``!s``, ``!r``, and ``!a`` are supported by this PEP in order
to minimize the differences with ``str.format()``. ``!s``, ``!r``, and
``!a`` are required in ``str.format()`` because it does not allow the
execution of arbitrary expressions.

Lambdas inside expressions
--------------------------

Because lambdas use the ``':'`` character, they cannot appear outside
of parentheses in an expression. The colon is interpreted as the start
of the format specifier, which means the start of the lambda
expression is seen and is syntactically invalid. As there's no
practical use for a plain lambda in an f-string expression, this is
not seen as much of a limitation.

If you feel you must use lambdas, they may be used inside of parentheses::

  >>> f'{(lambda x: x*2)(3)}'
  '6'

Can't combine with 'u'
--------------------------

The 'u' prefix was added to Python 3.3 in PEP 414 as a means to ease
source compatibility with Python 2.7. Because Python 2.7 will never
support f-strings, there is nothing to be gained by being able to
combine the 'f' prefix with 'u'.

Examples from Python's source code
==================================

Here are some examples from Python source code that currently use
``str.format()``, and how they would look with f-strings. This PEP
does not recommend wholesale converting to f-strings, these are just
examples of real-world usages of ``str.format()`` and how they'd look
if written from scratch using f-strings.

``Lib/asyncio/locks.py``::

  extra = '{},waiters:{}'.format(extra, len(self._waiters))
  extra = f'{extra},waiters:{len(self._waiters)}'

``Lib/configparser.py``::

  message.append(" [line {0:2d}]".format(lineno))
  message.append(f" [line {lineno:2d}]")

``Tools/clinic/clinic.py``::

  methoddef_name = "{}_METHODDEF".format(c_basename.upper())
  methoddef_name = f"{c_basename.upper()}_METHODDEF"

``python-config.py``::

  print("Usage: {0} [{1}]".format(sys.argv[0], '|'.join('--'+opt for opt in valid_opts)), file=sys.stderr)
  print(f"Usage: {sys.argv[0]} [{'|'.join('--'+opt for opt in valid_opts)}]", file=sys.stderr)

References
==========

.. [#] %-formatting
       (https://docs.python.org/3/library/stdtypes.html#printf-style-string-formatting)

.. [#] str.format
       (https://docs.python.org/3/library/string.html#formatstrings)

.. [#] string.Template documentation
       (https://docs.python.org/3/library/string.html#template-strings)

.. [#] Formatting using locals() and globals()
       (https://mail.python.org/pipermail/python-ideas/2015-July/034671.html)

.. [#] Avoid locals() and globals()
       (https://mail.python.org/pipermail/python-ideas/2015-July/034701.html)

.. [#] String literal description
       (https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals)

.. [#] ast.parse() documentation
       (https://docs.python.org/3/library/ast.html#ast.parse)

.. [#] Start of python-ideas discussion
       (https://mail.python.org/pipermail/python-ideas/2015-July/034657.html)

.. [#] Wikipedia article on string interpolation
       (https://en.wikipedia.org/wiki/String_interpolation)

.. [#] Differences in str.format() and f-string expressions
       (https://mail.python.org/pipermail/python-ideas/2015-July/034726.html)

.. [#] PEP 461 rejects bytes.format()
       (https://www.python.org/dev/peps/pep-0461/#proposed-variations)

Copyright
=========

This document has been placed in the public domain.


..
   Local Variables:
   mode: indented-text
   indent-tabs-mode: nil
   sentence-end-double-space: t
   fill-column: 70
   coding: utf-8
   End:

**Comment by anijjar** [comment]
PEP: 1
Title: PEP Purpose and Guidelines
Author: Barry Warsaw, Jeremy Hylton, David Goodger, Nick Coghlan
Status: Active
Type: Process
Content-Type: text/x-rst
Created: 13-Jun-2000
Post-History: 21-Mar-2001, 29-Jul-2002, 03-May-2003, 05-May-2012,
              07-Apr-2013


What is a PEP?
==============

PEP stands for Python Enhancement Proposal.  A PEP is a design
document providing information to the Python community, or describing
a new feature for Python or its processes or environment.  The PEP
should provide a concise technical specification of the feature and a
rationale for the feature.

We intend PEPs to be the primary mechanisms for proposing major new
features, for collecting community input on an issue, and for
documenting the design decisions that have gone into Python.  The PEP
author is responsible for building consensus within the community and
documenting dissenting opinions.

Because the PEPs are maintained as text files in a versioned
repository, their revision history is the historical record of the
feature proposal [1]_.


PEP Audience
============

The typical primary audience for PEPs are the core developers of the CPython
reference interpreter and their elected Steering Council, as well as developers
of other implementations of the Python language specification.

However, other parts of the Python community may also choose to use the process
(particularly for Informational PEPs) to document expected API conventions and
to manage complex design coordination problems that require collaboration across
multiple projects.


PEP Types
=========

There are three kinds of PEP:

1. A **Standards Track** PEP describes a new feature or implementation
   for Python. It may also describe an interoperability standard that will
   be supported outside the standard library for current Python versions
   before a subsequent PEP adds standard library support in a future
   version.

2. An **Informational** PEP describes a Python design issue, or
   provides general guidelines or information to the Python community,
   but does not propose a new feature.  Informational PEPs do not
   necessarily represent a Python community consensus or
   recommendation, so users and implementers are free to ignore
   Informational PEPs or follow their advice.

3. A **Process** PEP describes a process surrounding Python, or
   proposes a change to (or an event in) a process.  Process PEPs are
   like Standards Track PEPs but apply to areas other than the Python
   language itself.  They may propose an implementation, but not to
   Python's codebase; they often require community consensus; unlike
   Informational PEPs, they are more than recommendations, and users
   are typically not free to ignore them.  Examples include
   procedures, guidelines, changes to the decision-making process, and
   changes to the tools or environment used in Python development.
   Any meta-PEP is also considered a Process PEP.


PEP Workflow
============

Python's Steering Council
-------------------------

There are several references in this PEP to the "Steering Council" or "Council".
This refers to the current members of the elected Steering Council described
in PEP 13 [5]_, in their role as the final authorities on whether or not PEPs
will be accepted or rejected.


Python's Core Developers
------------------------

There are several references in this PEP to "core developers". This refers to
the currently active Python core team members described in PEP 13 [5]_.


Python's BDFL
-------------

This PEP still uses the title "BDFL-Delegate" for PEP decision makers. This is
a historical reference to Python's previous governance model, where all design
authority ultimately derived from Guido van Rossum, the original creator of the
Python programming language. By contrast, the Steering Council's design
authority derives from their election by the currently active core developers.


PEP Editors
-----------

The PEP editors are individuals responsible for managing the administrative
and editorial aspects of the PEP workflow (e.g. assigning PEP numbers and
changing their status).  See `PEP Editor Responsibilities & Workflow`_ for
details.

PEP editorship is by invitation of the current editors, and they can be
contacted via the address <peps@python.org>, but you may only need to use this
to contact the editors semi-privately.  All of the PEP workflow can be
conducted via the GitHub `PEP repository`_ issues and pull requests.


Start with an idea for Python
-----------------------------

The PEP process begins with a new idea for Python.  It is highly
recommended that a single PEP contain a single key proposal or new
idea. Small enhancements or patches often don't need
a PEP and can be injected into the Python development workflow with a
patch submission to the Python `issue tracker`_. The more focused the
PEP, the more successful it tends to be.  The PEP editors reserve the
right to reject PEP proposals if they appear too unfocused or too
broad.  If in doubt, split your PEP into several well-focused ones.

Each PEP must have a champion -- someone who writes the PEP using the style
and format described below, shepherds the discussions in the appropriate
forums, and attempts to build community consensus around the idea.  The PEP
champion (a.k.a. Author) should first attempt to ascertain whether the idea is
PEP-able.  Posting to the comp.lang.python newsgroup
(a.k.a. python-list@python.org mailing list) or the python-ideas@python.org
mailing list is the best way to go about this.

Vetting an idea publicly before going as far as writing a PEP is meant
to save the potential author time. Many ideas have been brought
forward for changing Python that have been rejected for various
reasons. Asking the Python community first if an idea is original
helps prevent too much time being spent on something that is
guaranteed to be rejected based on prior discussions (searching
the internet does not always do the trick). It also helps to make sure
the idea is applicable to the entire community and not just the author.
Just because an idea sounds good to the author does not
mean it will work for most people in most areas where Python is used.

Once the champion has asked the Python community as to whether an
idea has any chance of acceptance, a draft PEP should be presented to
python-ideas.  This gives the author a chance to flesh out the draft
PEP to make properly formatted, of high quality, and to address
initial concerns about the proposal.


Submitting a PEP
----------------

Following a discussion on python-ideas, the workflow varies based on whether
any of the PEP's co-authors are core developers. If one or more of the PEP's
co-authors are core developers, they are responsible for following the process
outlined below. Otherwise (i.e. none of the co-authors are core developers),
then the PEP author(s) will need to find a sponsor for the PEP.

Ideally, a core developer sponsor is identified, but non-core sponsors may also
be selected with the approval of the Steering Council.  The sponsor's job is to
provide guidance to the PEP author to help them through the logistics of the
PEP process (somewhat acting like a mentor).  Being a sponsor does **not**
disqualify that person from becoming a co-author or BDFL-Delegate later on (but
not both).  The sponsor of a PEP is recorded in the "Sponsor:" field of the
header.

Once the sponsor or the core developer(s) co-authoring the PEP deem the PEP
ready for submission, the proposal should be submitted as a draft PEP via a
`GitHub pull request`_.  The draft must be written in PEP style as described
below, else it will fail review immediately (although minor errors may be
corrected by the editors).

The standard PEP workflow is:

* You, the PEP author, fork the `PEP repository`_, and create a file named
  ``pep-9999.rst`` that contains your new PEP.  Use "9999" as your draft PEP
  number.

* In the "Type:" header field, enter "Standards Track",
  "Informational", or "Process" as appropriate, and for the "Status:"
  field enter "Draft".  For full details, see `PEP Header Preamble`_.

* Push this to your GitHub fork and submit a pull request.

* The PEP editors review your PR for structure, formatting, and other
  errors.  For a reST-formatted PEP, PEP 12 is provided as a template.
  It also provides a complete introduction to reST markup that is used
  in PEPs.  Approval criteria are:

  * It sound and complete.  The ideas must make technical sense.  The
    editors do not consider whether they seem likely to be accepted.
  * The title accurately describes the content.
  * The PEP's language (spelling, grammar, sentence structure, etc.)
    and code style (examples should match PEP 8 & PEP 7) should be
    correct and conformant.  The PEP will be checked for formatting
    (plain text or reStructuredText) by Travis CI, and will not be
    approved until this passes.

  Editors are generally quite lenient about this initial review,
  expecting that problems will be corrected by the reviewing process.
  **Note:** Approval of the PEP is no guarantee that there are no
  embarrassing mistakes!  Correctness is the responsibility of authors
  and reviewers, not the editors.

  If the PEP isn't ready for approval, an editor will send it back to
  the author for revision, with specific instructions.

* Once approved, they will assign your PEP a number.

Once the review process is complete, and the PEP editors approve it (note that
this is *not* the same as accepting your PEP!), they will squash commit your
pull request onto master.

The PEP editors will not unreasonably deny publication of a PEP.  Reasons for
denying PEP status include duplication of effort, being technically unsound,
not providing proper motivation or addressing backwards compatibility, or not
in keeping with the Python philosophy.  The Steering Council can be consulted
during the approval phase, and are the final arbiter of a draft's PEP-ability.

Developers with git push privileges for the `PEP repository`_ may claim PEP
numbers directly by creating and committing a new PEP. When doing so, the
developer must handle the tasks that would normally be taken care of by the
PEP editors (see `PEP Editor Responsibilities & Workflow`_). This includes
ensuring the initial version meets the expected standards for submitting a
PEP. Alternately, even developers may choose to submit PEPs via pull request.
When doing so, let the PEP editors know you have git push privileges and they
can guide you through the process of updating the PEP repository directly.

As updates are necessary, the PEP author can check in new versions if they
(or a collaborating developer) have git push privileges.

After a PEP number has been assigned, a draft PEP may be discussed further on
python-ideas (getting a PEP number assigned early can be useful for ease of
reference, especially when multiple draft PEPs are being considered at the
same time). Eventually, all Standards Track PEPs must  be sent to the
`python-dev list <mailto:python-dev@python.org>`__ for review as described
in the next section.

Standards Track PEPs consist of two parts, a design document and a
reference implementation.  It is generally recommended that at least a
prototype implementation be co-developed with the PEP, as ideas that sound
good in principle sometimes turn out to be impractical when subjected to the
test of implementation.

PEP authors are responsible for collecting community feedback on a PEP
before submitting it for review. However, wherever possible, long
open-ended discussions on public mailing lists should be avoided.
Strategies to keep the discussions efficient include: setting up a
separate SIG mailing list for the topic, having the PEP author accept
private comments in the early design phases, setting up a wiki page, etc.
PEP authors should use their discretion here.


PEP Review & Resolution
-----------------------

Once the authors have completed a PEP, they may request a review for
style and consistency from the PEP editors.

However, content review and final acceptance of the PEP must be requested of the
core developers, usually via an email to the python-dev mailing list.

To expedite the process in selected cases (e.g. when a change is clearly
beneficial and ready to be accepted, but the PEP hasn't been formally submitted
for review yet), the Steering Council may also initiate a PEP review, first
notifying the PEP author(s) and giving them a chance to make revisions.

The final authority for PEP approval is the Steering Council. However, whenever
a new PEP is put forward, any core developer that believes they are suitably
experienced to make the final decision on that PEP may offer to serve as
the BDFL-Delegate for that PEP, and they will then have the authority to approve
(or reject) that PEP. Individuals taking on this responsibility are free to seek
additional guidance from the Steering Council at any time, and are also expected
to take the advice and perspectives of other core developers into account.

The designated decision maker for each PEP is recorded in the "BDFL-Delegate"
header in the PEP.

Such self-nominations are accepted by default, but may be explicitly declined by
the Steering Council. Possible reasons for the Steering Council declining a
self-nomination as BDFL-Delegate include, but are not limited to, perceptions of
a potential conflict of interest (e.g. working for the same organisation as the
PEP submitter), or simply considering another potential BDFL-Delegate to be
more appropriate. If core developers (or other community members) have concerns
regarding the suitability of a BDFL-Delegate for any given PEP, they may ask
the Steering Council to review the delegation.

If no volunteer steps forward, then the Steering Council will approach core
developers (and potentially other Python community members) with relevant
expertise, in an attempt to identify a candidate that is willing to serve as
BDFL-Delegate for that PEP. If no suitable candidate can be found, then the
PEP will be marked as Deferred until one is available.

Previously appointed BDFL-Delegates may choose to step down, or be asked to step
down by the Council, in which case a new BDFL-Delegate will be appointed in the
same manner as for a new PEP (including deferral of the PEP if no suitable
replacement can be found). In the event that a BDFL-Delegate is asked to step
down, this will overrule any prior acceptance or rejection of the PEP, and it
will revert to Draft status.

With the approval of the Steering Council, PEP review and resolution may also
occur on a list other than python-dev (for example, distutils-sig for packaging
related PEPs that don't immediately affect the standard library). In these
cases, the "Discussions-To" heading in the PEP will identify the appropriate
alternative list where discussion, review and pronouncement on the PEP will
occur.

When such standing delegations are put in place, the Steering Council will
maintain sufficient public records to allow subsequent Councils, the core
developers, and the wider Python community to understand the delegations that
currently exist, why they were put in place, and the circumstances under which
they may no longer be needed.

For a PEP to be accepted it must meet certain minimum criteria.  It
must be a clear and complete description of the proposed enhancement.
The enhancement must represent a net improvement.  The proposed
implementation, if applicable, must be solid and must not complicate
the interpreter unduly.  Finally, a proposed enhancement must be
"pythonic" in order to be accepted by the Steering Council.  (However,
"pythonic" is an imprecise term; it may be defined as whatever is acceptable to
the Steering Council.  This logic is intentionally circular.)  See PEP 2 [2]_
for standard library module acceptance criteria.

Once a PEP has been accepted, the reference implementation must be
completed.  When the reference implementation is complete and incorporated
into the main source code repository, the status will be changed to "Final".

To allow gathering of additional design and interface feedback before committing
to long term stability for a language feature or standard library API, a PEP
may also be marked as "Provisional". This is short for "Provisionally Accepted",
and indicates that the proposal has been accepted for inclusion in the reference
implementation, but additional user feedback is needed before the full design
can be considered "Final". Unlike regular accepted PEPs, provisionally accepted
PEPs may still be Rejected or Withdrawn *even after the related changes have
been included in a Python release*.

Wherever possible, it is considered preferable to reduce the scope of a proposal
to avoid the need to rely on the "Provisional" status (e.g. by deferring some
features to later PEPs), as this status can lead to version compatibility
challenges in the wider Python ecosystem. PEP 411 provides additional details
on potential use cases for the Provisional status.

A PEP can also be assigned the status "Deferred".  The PEP author or an
editor can assign the PEP this status when no progress is being made
on the PEP.  Once a PEP is deferred, a PEP editor can re-assign it
to draft status.

A PEP can also be "Rejected".  Perhaps after all is said and done it
was not a good idea.  It is still important to have a record of this
fact. The "Withdrawn" status is similar - it means that the PEP author
themselves has decided that the PEP is actually a bad idea, or has
accepted that a competing proposal is a better alternative.

When a PEP is Accepted, Rejected or Withdrawn, the PEP should be updated
accordingly. In addition to updating the status field, at the very least
the Resolution header should be added with a link to the relevant post
in the python-dev mailing list archives.

PEPs can also be superseded by a different PEP, rendering the original
obsolete.  This is intended for Informational PEPs, where version 2 of
an API can replace version 1.

The possible paths of the status of PEPs are as follows:

.. image:: pep-0001-process_flow.png
   :alt: PEP process flow diagram

While not shown in the diagram, "Accepted" PEPs may technically move to
"Rejected" or "Withdrawn" even after acceptance. This will only occur if
the implementation process reveals fundamental flaws in the design that were
not noticed prior to acceptance of the PEP. Unlike Provisional PEPs, these
transitions are only permitted if the accepted proposal has *not* been included
in a Python release - released changes must instead go through the regular
deprecation process (which may require a new PEP providing the rationale for
the deprecation).

Some Informational and Process PEPs may also have a status of "Active"
if they are never meant to be completed.  E.g. PEP 1 (this PEP).


PEP Maintenance
---------------

In general, Standards track PEPs are no longer modified after they have
reached the Final state. Once a PEP has been completed, the Language and
Standard Library References become the formal documentation of the expected
behavior.

If changes based on implementation experience and user feedback are made to
Standards track PEPs while in the Accepted or Provisional State, those changes
should be noted in the PEP, such that the PEP accurately describes the state of
the implementation at the point where it is marked Final.

Informational and Process PEPs may be updated over time to reflect changes
to development practices and other details. The precise process followed in
these cases will depend on the nature and purpose of the PEP being updated.



What belongs in a successful PEP?
=================================

Each PEP should have the following parts/sections:

1. Preamble -- RFC 822 style headers containing meta-data about the
   PEP, including the PEP number, a short descriptive title (limited
   to a maximum of 44 characters), the names, and optionally the
   contact info for each author, etc.

2. Abstract -- a short (~200 word) description of the technical issue
   being addressed.

3. Motivation -- The motivation is critical for PEPs that want to
   change the Python language, library, or ecosystem.  It should
   clearly explain why the existing language specification is
   inadequate to address the problem that the PEP solves.  PEP
   submissions without sufficient motivation may be rejected outright.

4. Rationale -- The rationale fleshes out the specification by
   describing why particular design decisions were made.  It should
   describe alternate designs that were considered and related work,
   e.g. how the feature is supported in other languages.

   The rationale should provide evidence of consensus within the
   community and discuss important objections or concerns raised
   during discussion.

5. Specification -- The technical specification should describe the
   syntax and semantics of any new language feature.  The
   specification should be detailed enough to allow competing,
   interoperable implementations for at least the current major Python
   platforms (CPython, Jython, IronPython, PyPy).

6. Backwards Compatibility -- All PEPs that introduce backwards
   incompatibilities must include a section describing these
   incompatibilities and their severity.  The PEP must explain how the
   author proposes to deal with these incompatibilities.  PEP
   submissions without a sufficient backwards compatibility treatise
   may be rejected outright.

7. Security Implications -- If there are security concerns in relation
   to the PEP, those concerns should be explicitly written out to make
   sure reviewers of the PEP are aware of them.

8. How to Teach This -- For a PEP that adds new functionality or changes
   language behavior, it is helpful to include a section on how to
   teach users, new and experienced, how to apply the PEP to their
   work.

   This section may include key points and recommended documentation
   changes that would help users adopt a new feature or migrate their
   code to use a language change.

9. Reference Implementation -- The reference implementation must be
   completed before any PEP is given status "Final", but it need not
   be completed before the PEP is accepted.  While there is merit
   to the approach of reaching consensus on the specification and
   rationale before writing code, the principle of "rough consensus
   and running code" is still useful when it comes to resolving many
   discussions of API details.

   The final implementation must include test code and documentation
   appropriate for either the Python language reference or the
   standard library reference.

10. Rejected Ideas -- Throughout the discussion of a PEP, various ideas
    will be proposed which are not accepted. Those rejected ideas should
    be recorded along with the reasoning as to why they were rejected.
    This both helps record the thought process behind the final version
    of the PEP as well as preventing people from bringing up the same
    rejected idea again in subsequent discussions.

    In a way this section can be thought of as a breakout section of the
    Rationale section that is focused specifically on why certain ideas
    were not ultimately pursued.

11. Open Issues -- While a PEP is in draft, ideas can come up which
    warrant further discussion. Those ideas should be recorded so people
    know that they are being thought about but do not have a concrete
    resolution. This helps make sure all issues required for the PEP to be
    ready for consideration are complete complete and reduces people
    duplicating prior discussion.

12. References -- A collection of URLs used as references through the PEP.

13. Copyright/license -- Each new PEP must be placed under a dual license of
    public domain and CC0-1.0-Universal_ (see this PEP for an example).


PEP Formats and Templates
=========================

PEPs are UTF-8 encoded text files using the reStructuredText_ format.
ReStructuredText_ allows for rich markup that is still quite easy to
read, but also results in good-looking and functional HTML. PEP 12
contains instructions and a template [4]_ for reStructuredText PEPs.

The PEP text files are automatically converted to HTML [6]_ for easier
`online reading <https://www.python.org/dev/peps/>`__.


PEP Header Preamble
===================

Each PEP must begin with an RFC 822 style header preamble.  The headers
must appear in the following order.  Headers marked with "*" are
optional and are described below.  All other headers are required. ::

    PEP: <pep number>
    Title: <pep title>
    Author: <list of authors' real names and optionally, email addrs>
  * Sponsor: <real name of sponsor>
  * BDFL-Delegate: <PEP czar's real name>
  * Discussions-To: <email address>
    Status: <Draft | Active | Accepted | Provisional | Deferred | Rejected |
             Withdrawn | Final | Superseded>
    Type: <Standards Track | Informational | Process>
  * Content-Type: <text/x-rst | text/plain>
  * Requires: <pep numbers>
    Created: <date created on, in dd-mmm-yyyy format>
  * Python-Version: <version number>
    Post-History: <dates of postings to python-ideas and/or python-dev>
  * Replaces: <pep number>
  * Superseded-By: <pep number>
  * Resolution: <url>

The Author header lists the names, and optionally the email addresses
of all the authors/owners of the PEP.  The format of the Author header
value must be

    Random J. User <address@dom.ain>

if the email address is included, and just

    Random J. User

if the address is not given.  For historical reasons the format
"address@dom.ain (Random J. User)" may appear in a PEP, however new
PEPs must use the mandated format above, and it is acceptable to
change to this format when PEPs are updated.

If there are multiple authors, each should be on a separate line
following RFC 2822 continuation line conventions.  Note that personal
email addresses in PEPs will be obscured as a defense against spam
harvesters.

The Sponsor field records which developer (core, or otherwise approved by the
Steering Council) is sponsoring the PEP.  If one of the authors of the PEP is a
core developer then no sponsor is necessary and thus this field should be left
out.

The BDFL-Delegate field is used to record the individual appointed by the
Steering Council to make the final decision on whether or not to approve or
reject a PEP. (The delegate's email address is currently omitted due to a
limitation in the email address masking for reStructuredText PEPs)

*Note: The Resolution header is required for Standards Track PEPs
only.  It contains a URL that should point to an email message or
other web resource where the pronouncement about the PEP is made.*

For a PEP where final pronouncement will be made on a list other than
python-dev, a Discussions-To header will indicate the mailing list
or URL where the pronouncement will occur. A temporary Discussions-To header
may also be used when a draft PEP is being discussed prior to submission for
pronouncement. No Discussions-To header is necessary if the PEP is being
discussed privately with the author, or on the python-list, python-ideas
or python-dev mailing lists.  Note that email addresses in the
Discussions-To header will not be obscured.

The Type header specifies the type of PEP: Standards Track,
Informational, or Process.

The format of a PEP is specified with a Content-Type header.  The
acceptable values are "text/plain" for plaintext PEPs (see PEP 9 [3]_)
and "text/x-rst" for reStructuredText PEPs (see PEP 12 [4]_).
reStructuredText is strongly preferred, but for backwards
compatibility plain text is currently still the default if no
Content-Type header is present.

The Created header records the date that the PEP was assigned a
number, while Post-History is used to record the dates of when new
versions of the PEP are posted to python-ideas and/or python-dev.  Both
headers should be in dd-mmm-yyyy format, e.g. 14-Aug-2001.

Standards Track PEPs will typically have a Python-Version header which
indicates the version of Python that the feature will be released with.
Standards Track PEPs without a Python-Version header indicate
interoperability standards that will initially be supported through
external libraries and tools, and then potentially supplemented by a later PEP
to add support to the standard library. Informational and Process PEPs do
not need a Python-Version header.

PEPs may have a Requires header, indicating the PEP numbers that this
PEP depends on.

PEPs may also have a Superseded-By header indicating that a PEP has
been rendered obsolete by a later document; the value is the number of
the PEP that replaces the current document.  The newer PEP must have a
Replaces header containing the number of the PEP that it rendered
obsolete.


Auxiliary Files
===============

PEPs may include auxiliary files such as diagrams.  Such files should be
named ``pep-XXXX-Y.ext``, where "XXXX" is the PEP number, "Y" is a
serial number (starting at 1), and "ext" is replaced by the actual
file extension (e.g. "png").

Alternatively, all support files may be placed in a subdirectory called
``pep-XXXX``, where "XXXX" is the PEP number. When using a subdirectory, there
are no constraints on the names used in files.


Reporting PEP Bugs, or Submitting PEP Updates
=============================================

How you report a bug, or submit a PEP update depends on several
factors, such as the maturity of the PEP, the preferences of the PEP
author, and the nature of your comments.  For the early draft stages
of the PEP, it's probably best to send your comments and changes
directly to the PEP author.  For more mature, or finished PEPs you may
want to submit corrections as a `GitHub issue`_ or `GitHub pull request`_ so that
your changes don't get lost.

When in doubt about where to send your changes, please check first
with the PEP author and/or a PEP editor.

PEP authors with git push privileges for the PEP repository can update the
PEPs themselves by using "git push" or the GitHub PR interface to submit their
changes.


Transferring PEP Ownership
==========================

It occasionally becomes necessary to transfer ownership of PEPs to a
new champion.  In general, it is preferable to retain the original author as
a co-author of the transferred PEP, but that's really up to the
original author.  A good reason to transfer ownership is because the
original author no longer has the time or interest in updating it or
following through with the PEP process, or has fallen off the face of
the 'net (i.e. is unreachable or not responding to email).  A bad
reason to transfer ownership is because the author doesn't agree with the
direction of the PEP.  One aim of the PEP process is to try to build
consensus around a PEP, but if that's not possible, an author can always
submit a competing PEP.

If you are interested in assuming ownership of a PEP, you can also do this via
pull request.  Fork the `PEP repository`_, make your ownership modification,
and submit a pull request.  You should also send a message asking to take
over, addressed to both the original author and the PEP editors
<peps@python.org>.  If the original author doesn't respond to email in a
timely manner, the PEP editors will make a unilateral decision (it's not like
such decisions can't be reversed :).


PEP Editor Responsibilities & Workflow
======================================

A PEP editor must subscribe to the <peps@python.org> list and must watch the
`PEP repository`_.  Most correspondence regarding PEP administration can be
handled through GitHub issues and pull requests, but you may also use
<peps@python.org> for semi-private discussions.  Please do not cross-post!

For each new PEP that comes in an editor does the following:

* Make sure that the PEP is either co-authored by a core developer, has a core
  developer as a sponsor, or has a sponsor specifically approved for this PEP
  by the Steering Council.

* Read the PEP to check if it is ready: sound and complete.  The ideas
  must make technical sense, even if they don't seem likely to be
  accepted.

* The title should accurately describe the content.

* The file name extension is correct (i.e. ``.rst``).

* Skim the PEP for obvious defects in language (spelling, grammar,
  sentence structure, etc.), and code style (examples should conform to
  PEP 8 & PEP 7).  Editors may correct problems themselves, but are
  not required to do so.  (Text format is checked by Travis CI.)

If the PEP isn't ready, an editor will send it back to the author for
revision, with specific instructions.  If reST formatting is a
problem, ask the author(s) to use PEP 12 as a template and resubmit.

Once the PEP is ready for the repository, a PEP editor will:

* Assign a PEP number (almost always just the next available number,
  but sometimes it's a special/joke number, like 666 or 3141).
  (Clarification: For Python 3, numbers in the 3000s were used for
  Py3k-specific proposals.  But now that all new features go into
  Python 3 only, the process is back to using numbers in the 100s again.
  Remember that numbers below 100 are meta-PEPs.)

* Check that the author has correctly labeled the PEP's type
  ("Standards Track", "Informational", or "Process"), and marked its
  status as "Draft".

* Add the PEP to a local fork of the PEP repository.  For workflow
  instructions, follow `The Python Developers Guide <http://docs.python.org/devguide>`_

  The git repo for the peps is::

   https://github.com/python/peps

* Run ``./genpepindex.py`` and ``./pep2html.py <PEP Number>`` to ensure they
  are generated without errors. If either triggers errors, then the web site
  will not be updated to reflect the PEP changes.

* Commit and push the new (or updated) PEP

* Monitor python.org to make sure the PEP gets added to the site
  properly. If it fails to appear, running ``make`` will build all of the
  current PEPs. If any of these are triggering errors, they must be
  corrected before any PEP will update on the site.

* Send email back to the PEP author with next steps (post to
  python-list & -dev).

Updates to existing PEPs should be submitted as a `GitHub pull request`_.
Questions may of course still be sent to <peps@python.org>.

Many PEPs are written and maintained by developers with write access
to the Python codebase.  The PEP editors monitor the python-checkins
list for PEP changes, and correct any structure, grammar, spelling, or
markup mistakes they see.

PEP editors don't pass judgment on PEPs.  They merely do the
administrative & editorial part (which is generally a low volume task).

Resources:

* `Index of Python Enhancement Proposals <http://www.python.org/dev/peps/>`_

* `Following Python's Development
  <http://docs.python.org/devguide/communication.html>`_

* `Python Developer's Guide <http://docs.python.org/devguide/>`_

* `Frequently Asked Questions for Developers
  <http://docs.python.org/devguide/faq.html>`_


References and Footnotes
========================

.. [1] This historical record is available by the normal git commands
   for retrieving older revisions, and can also be browsed via HTTP here:
   https://github.com/python/peps

.. [2] PEP 2, Procedure for Adding New Modules
   (http://www.python.org/dev/peps/pep-0002)

.. [3] PEP 9, Sample Plaintext PEP Template
   (http://www.python.org/dev/peps/pep-0009)

.. [4] PEP 12, Sample reStructuredText PEP Template
   (http://www.python.org/dev/peps/pep-0012)

.. [5] PEP 13, Python Language Governance
   (http://www.python.org/dev/peps/pep-0013)

.. [6] More details on the PEP rendering and publication process can be found
   in the PEPs repo README at
   https://github.com/python/peps/blob/master/README.rst

.. _issue tracker:
   http://bugs.python.org/

.. _CC0-1.0-Universal: https://choosealicense.com/licenses/cc0-1.0/

.. _reStructuredText: http://docutils.sourceforge.net/rst.html

.. _Docutils: http://docutils.sourceforge.net/

.. _PEP repository: https://github.com/python/peps

.. _`GitHub pull request`: https://github.com/python/peps/pulls

.. _`GitHub issue`: https://github.com/python/peps/issues


Copyright
=========

This document is placed in the public domain or under the
CC0-1.0-Universal license, whichever is more permissive.


..
   Local Variables:
   mode: indented-text
   indent-tabs-mode: nil
   sentence-end-double-space: t
   fill-column: 70
   coding: utf-8
   End:
